torch/csrc/distributed/rpc/rref_impl.cpp

```diff
@@ -250,12 +251,12 @@ void OwnerRRef::setValue(IValue&& value) {
 void OwnerRRef::setError(const std::string& error) {
   std::unique_lock<std::mutex> lock(mutex_);
   error_ = error;
-  std::shared_ptr<FutureIValue> future;
+  std::shared_ptr<JitFuture> future;
   future.swap(future_);
   lock.unlock();
   valueCV_.notify_all();
-  if (future.get()) {
-    future->setErrorIfNeeded(error);
+  if (future.get() && !future->completed()) {
```

Comment:
`setErrorIfNeeded` is used for the case to handle when future completed right?

Reply:
It's used when there are multiple threads trying to set the future. An example in RpcAgent is that there could be 1) a timeout watch dog thread and 2) a business logic callback handler running into a failure. They both want to set error on the future.

Reply:
@wanchaol 
I added setErrorIfNeeded to `c10::ivalue::Future`, because this 2 calls don't not work atomically.

assert `!future->completed()` +  `future->setError(error)` must be atomic.